### PR TITLE
Add OpenSearchDocument XML (#5399)

### PR DESCRIPTION
**Pull request by RoiEXLab**
This adresses #5399
This generates an xml file into the build folder (accessible using `/assets/osd.xml`) every time nodebb is started, based on the current desrciption and title settings...
Not sure, if all settings are retrieved correctly though (nconf url for example)

### CODE REVIEW

**Comment by RoiEXLab** [comment]
Wow. Your tests are really picky!

btw... npm test seems to be broken on windows... `212292 problems (212288 errors, 4 warnings)` doesn't seem to be right^^

**Comment by RoiEXLab** [comment]
Just in case you were wondering: the xml dependency is already required by rss, but travis didn't seemed to like that and errored only in [Job \#8040.4](https://travis-ci.org/NodeBB/NodeBB/jobs/207902798).
Thats why I added it to the package.json

**Comment by pitaj** [comment]
This is probably because your global git setting downloads as crlf. 

Adding 
```
* text eol=lf
* text eol=lf
```
To the `.gitattributes` file would likely fix this.

**Comment by RoiEXLab** [comment]
@pitaj tried that and every possible other combination, and it isn't working...
Doesn't matter though... I won't use that feature too often ^^

**Comment by pitaj** [comment]
Well doing it after you've cloned the repo don't change anything, but for other people it could help.

**Comment by RoiEXLab** [comment]
@pitaj Ahh I see... It seems that `core.autocrlf` is enabled on my machine... that's probably why

**Comment by julianlam** [comment]
Whoa, hooking in the build step is overkill here, since the file almost never changes.

Defining an express route and sending in the html on request is much lighter. Keeping `src/meta/osd.js` is fine though 😄

**Comment by RoiEXLab** [comment]
I added in the build step, because the file almost never changes ^^
This way the file is just located along with the other resources and doesn't need to be "recalculated" every time a new chromium based client opens the website.
A production server shouldn't restart too often anyways

**Comment by RoiEXLab** [comment]
@julianlam what should I do now?
Should I keep this file accessible at /assets/osd.xml or should I move it to the sitemap code?

**Comment by RoiEXLab** [comment]
@julienlam fixed the code, moved the osd.js file to the controllers directory.
The xml file is now generated dynamically!

I don't get how coveralls works... Any ideas on how to improve the score?

**Comment by julianlam** [comment]
Don't worry about coveralls 😆

File: src/controllers/index.js [hunk 1]
```diff
@@ -32,6 +32,7 @@ Controllers.mods = require('./mods');
 Controllers.sitemap = require('./sitemap');
 Controllers['404'] = require('./404');
 Controllers.errors = require('./errors');
+Controllers.handleOSDRequest = require('./osd');
```
Comment (julianlam) [inline src/controllers/index.js]: To conform, I'd move this up two lines (for 404 and error are at the bottom), and change the property name to `osd`

**Comment by RoiEXLab** [comment]
@julianlam ok ^^
The xml is now served at `/osd.xml` in order to not interfere with the assets stuff btw.

File: src/controllers/osd.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+var plugins = require('../plugins');
+var meta = require('../meta');
+
+module.exports = function (req, res, next) {
+	if (plugins.hasListeners('filter:search.query')) {
+		res.type('application/xml').send(generateXML());
+	} else {
+		next();
+	}
+};
```
Comment (julianlam) [inline src/controllers/osd.js]: Please have the function export an object with the middleware above as a property of said object. I'd recommend `handle` as the method name, so below, you can `app.get('/osd.xml', controllers.osd.handle);`

File: src/routes/meta.js [hunk 3]
```diff
@@ -8,4 +8,5 @@ module.exports = function (app, middleware, controllers) {
 	app.get('/robots.txt', controllers.robots);
 	app.get('/manifest.json', controllers.manifest);
 	app.get('/css/previews/:theme', controllers.admin.themes.get);
+	app.get('/osd.xml', controllers.handleOSDRequest);
```
Comment (julianlam) [inline src/routes/meta.js]: As above, `controllers.osd.handle`

**Comment by julianlam** [comment]
This PR does not add the `<link>` tag right?

**Comment by RoiEXLab** [comment]
@julianlam It does!

**Comment by RoiEXLab** [comment]
@julianlam Edited the files...

**Comment by RoiEXLab** [comment]
if no search plugin is installed, `/osd.xml` returns a 404 error and no link tag is created to avoid other 404 errors

**Pull request merged**